### PR TITLE
Guard against reflection failure when retrieving a pipe name for VBCSCompiler

### LINKED ISSUE #13966: NRE in build task
"E:\git\temp\temp.csproj" (default target) (1) ->
(CoreCompile target) ->
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error MSB3883: Unexpected exception:  [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : System.NullReferenceException: Object reference not set to an instance of an object. [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.GetUserName() [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.GetPipeNameForPath(String compilerExeDirectory) [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.RunServerCompilation(RequestLanguage language, List`1 arguments, BuildPaths buildPaths, String keepAlive, String libEnvVariable, CancellationToken cancellationToken) [E:\git\temp\temp.csproj]
  E:\git\cli-1\artifacts\win10-x64\stage2\sdk\1.0.0-preview3-003633\Microsoft.CSharp.Core.targets(71,5): error : at Microsoft.CodeAnalysis.BuildTasks.ManagedCompiler.ExecuteTool(String pathToTool, String responseFileCommands, String commandLineCommands) [E:\git\temp\temp.csproj]

### PR DESCRIPTION
**Pull request by agocke**
Fixes #13966

/cc @jaredpar @dotnet/roslyn-compiler

### CODE REVIEW

File: src/Compilers/Shared/DesktopBuildClient.cs [hunk 1]
```diff
@@ -488,10 +495,13 @@ private static bool CheckPipeConnectionOwnership(NamedPipeClientStream pipeStrea
...
         {
-            return GetPipeNameForPath(buildPaths.ClientDirectory);
+            return GetPipeNameForPathOpt(buildPaths.ClientDirectory);
         }
 
-        internal static string GetPipeNameForPath(string compilerExeDirectory)
+        /// <returns>
+        /// Null if not enough information was found to create a valid pipe name.
+        /// </returns>
+        internal static string GetPipeNameForPathOpt(string compilerExeDirectory)
```
Comment (agocke) [inline src/Compilers/Shared/DesktopBuildClient.cs]: I usually use `TryGet` for out var + bool return, and Opt for something that may be null. The latter is very common in Roslyn. I don't really care either way, though.

File: src/Compilers/Shared/DesktopBuildClient.cs [hunk 2]
```diff
@@ -473,13 +480,13 @@ private static bool CheckPipeConnectionOwnership(NamedPipeClientStream pipeStrea
             (ObjectSecurity)typeof(PipeStream)
             .GetTypeInfo()
             .GetDeclaredMethod("GetAccessControl")
```
Comment (agocke) [inline src/Compilers/Shared/DesktopBuildClient.cs]: I think it's OK if this throws. The only usage is surrounded in a try-catch that swallows all exceptions and turns it into a failed build response.

File: src/Compilers/Shared/DesktopBuildClient.cs [hunk 3]
```diff
@@ -473,13 +480,13 @@ private static bool CheckPipeConnectionOwnership(NamedPipeClientStream pipeStrea
...
-            .Invoke(pipeStream, parameters: null);
+            ?.Invoke(pipeStream, parameters: null);
 
         private static string GetUserName() =>
             (string)typeof(Environment)
             .GetTypeInfo()
             .GetDeclaredProperty("UserName")
-            .GetMethod.Invoke(null, parameters: null);
+            ?.GetMethod?.Invoke(null, parameters: null);
```
Comment (agocke) [inline src/Compilers/Shared/DesktopBuildClient.cs]: I believe this can require certain security permissions in partial trust, but will not throw.

**Comment by agocke** [comment]
test windows_debug_unit32_prtest  please

**Comment by agocke** [comment]
test windows_debug_unit64_prtest please

**Comment by agocke** [comment]
ping @dotnet/roslyn-compiler One more review please

**Comment by AlekseyTs** [comment]
LGTM, modulo the lack of unit-tests.

**Comment by agocke** [comment]
@AlekseyTs Unfortunately, it's very difficult to cover this in a unit test, especially since it can only happen in .NET Core. I'm considering what options we have here.

**Pull request merged**